### PR TITLE
ScaLAPACKMatrix: copy functions

### PR DESCRIPTION
**Pull request by BenBrands**
May I ask you to have a look, @davydden?

### CODE REVIEW

File: include/deal.II/lac/scalapack.h [hunk 1]
```diff
@@ -179,6 +180,15 @@ class ScaLAPACKMatrix : protected TransposeTable<NumberType>
   ScaLAPACKMatrix<NumberType> &
   operator = (const FullMatrix<NumberType> &);
 
+  /**
+   * Copies the content of the locally owned @p matrix to the distributed matrix.
```
[the remark this answers is not in the document]
Comment (BenBrands) [reply include/deal.II/lac/scalapack.h]: I would prefer to have both options.
One where all processes have/get serial copy of the distributed matrix and one that needs only root to have/get the serial copy.
The first has to be implemented on a free evening :)
Certainly, we should adapt the naming of the routines. What do you think?

[the remark this answers is not in the document]
Comment (davydden) [reply include/deal.II/lac/scalapack.h]: > one that needs only root to have/get the serial copy.

do you mean that you do MPI communication to send this serial matrix where needed in distributed one?

**Comment by davydden** [comment]
@dealii/dealii 

AFAIK we don't have a similar user case in the library in vector and/or matrices, what @BenBrands wants to add is distribution/gathering of a serial dense matrix from/to some MPI process (root or given by rank) into the parallel dense ScaLAPACK matrix. What others think about this and how about naming methods as `ScaLAPACKMatrix::copy_from(const FullMatrix&, const unsigned int rank)` and `ScaLAPACKMatrix::copy_to(FullMatrix&, const unsigned int rank)` ? Obviously in both cases `FullMatrix` is only needed on the source/target process.

**Comment by BenBrands** [comment]
@masterleinad: I have to work on that a little more, but will not manage to do so in the next three or four weeks. When my schedule is more relaxed I will continue working on this PR.

**Comment by davydden** [comment]
@BenBrands I think the names in https://github.com/dealii/dealii/pull/6241#issuecomment-381887676 are fine, as far as I am concerned. And since nobody from @dealii/dealii is against that so far, I would just go for that structure/naming.

File: include/deal.II/lac/scalapack.h [hunk 2]
```diff
@@ -190,6 +191,16 @@ class ScaLAPACKMatrix : protected TransposeTable<NumberType>
   ScaLAPACKMatrix<NumberType> &
   operator=(const FullMatrix<NumberType> &);
 
+  /**
+   * Copies the content of the locally owned @p matrix to the distributed matrix.
+   * For all processes except the process with rank @p r the serial @p matrix is not referenced.
```
Comment (davydden) [inline include/deal.II/lac/scalapack.h]: Please renam ‘r’ to ‘rank’ (spell-out)

File: include/deal.II/lac/scalapack.h [hunk 3]
```diff
@@ -199,6 +210,15 @@ class ScaLAPACKMatrix : protected TransposeTable<NumberType>
   void
   copy_to(FullMatrix<NumberType> &matrix) const;
 
+  /**
+   * Copies the content of the distributed matrix into the locally replicated @p matrix
+   * on the process with rank @r. For all processes except @p r @p matrix is not referenced.
```
Comment (davydden) [inline include/deal.II/lac/scalapack.h]: Same with renaming of the argument

File: include/deal.II/lac/scalapack.h [hunk 4]
```diff
@@ -199,6 +210,15 @@ class ScaLAPACKMatrix : protected TransposeTable<NumberType>
...
+   * The distributed matrix and @p matrix on the process @p r must have matching dimensions.
+   *    * The user has to ensure that all processes call this with identical @p r.
```
Comment (davydden) [inline include/deal.II/lac/scalapack.h]: Extra ‘*’

File: source/lac/scalapack.cc [hunk 5]
```diff
@@ -263,6 +263,131 @@ ScaLAPACKMatrix<NumberType>::operator=(const FullMatrix<NumberType> &matrix)
...
+void
+ScaLAPACKMatrix<NumberType>::copy_from(const LAPACKFullMatrix<NumberType> &B,
+                                       const unsigned int                  r)
+{
+  if (n_rows * n_columns == 0)
+    return;
+
+  const unsigned int this_mpi_process(
+    Utilities::MPI::this_mpi_process(this->grid->mpi_communicator));
+
```
Comment (davydden) [inline source/lac/scalapack.cc]: Here and below you can add a check inside ‘#ifdef DEBUG’ that rank is the same everywhere. To that end u can get maximum over communicator via Utilities.

File: source/lac/scalapack.cc [hunk 6]
```diff
@@ -263,6 +263,131 @@ ScaLAPACKMatrix<NumberType>::operator=(const FullMatrix<NumberType> &matrix)
...
+  // root process has to be active in the grid of A
+  if (this_mpi_process == r)
+    {
+      Assert(grid->mpi_process_is_active, ExcInternalError());
+      Assert(n_rows == int(B.m()), ExcDimensionMismatch(n_rows, B.m()));
+      Assert(n_columns == int(B.n()), ExcDimensionMismatch(n_columns, B.n()));
+    }
+  // Create 1x1 grid for matrix B.
```
Comment (davydden) [inline source/lac/scalapack.cc]: Maybe add a comment that u copy via Scalapack functions and thus need grid.

And a brief summary of the algorithm

File: include/deal.II/lac/scalapack.h [hunk 7]
```diff
@@ -190,6 +191,17 @@ class ScaLAPACKMatrix : protected TransposeTable<NumberType>
   ScaLAPACKMatrix<NumberType> &
   operator=(const FullMatrix<NumberType> &);
 
+  /**
+   * Copies the content of the locally owned @p matrix to the distributed matrix.
+   * For all processes except the process with rank @p rank the serial @p matrix is not referenced.
+   * The user has to ensure that all processes call this with identical @p rank.
+   *
+   * The distributed matrix and @p matrix on process @p r must have matching dimensions.
```
Comment (davydden) [inline include/deal.II/lac/scalapack.h]: r -> rank

File: source/lac/scalapack.cc [hunk 8]
```diff
@@ -263,6 +263,141 @@ ScaLAPACKMatrix<NumberType>::operator=(const FullMatrix<NumberType> &matrix)
...
+  // This grid will be used to copy the serial matrix B to the distributed
+  // matrix using the ScaLAPACK routine pgemr2d.
+  MPI_Group group_A;
+  MPI_Comm_group(this->grid->mpi_communicator, &group_A);
+  const int              n = 1;
+  const std::vector<int> ranks(n, rank);
+  MPI_Group              group_B;
+  MPI_Group_incl(group_A, n, ranks.data(), &group_B);
+  MPI_Comm communicator_B;
+  MPI_Comm_create_group(MPI_COMM_WORLD, group_B, 0, &communicator_B);
```
Comment (davydden) [inline source/lac/scalapack.cc]: can we use `this->grid->mpi_communicator` here? We know that we copy to some rank within the communicator of this grid, right?

File: source/lac/scalapack.cc [hunk 9]
```diff
@@ -263,6 +263,141 @@ ScaLAPACKMatrix<NumberType>::operator=(const FullMatrix<NumberType> &matrix)
...
+                      &n_proc_rows_B,
+                      &n_proc_cols_B,
+                      &this_process_row_B,
+                      &this_process_column_B);
+      Assert(n_proc_rows_B * n_proc_cols_B == 1, ExcInternalError());
+      // the active process of grid B has to be process #rank of the
+      // communicator attached to A
+      Assert(this_mpi_process == rank, ExcInternalError());
+    }
+  bool mpi_process_is_active_B = false;
```
Comment (davydden) [inline source/lac/scalapack.cc]: oneliner:
`const bool mpi_process_is_active_B = (this_process_row_B >= 0 && this_process_column_B >= 0);`

File: source/lac/scalapack.cc [hunk 10]
```diff
@@ -263,6 +263,141 @@ ScaLAPACKMatrix<NumberType>::operator=(const FullMatrix<NumberType> &matrix)
...
+                &blacs_context_B,
+                &lda,
+                &info);
+      AssertThrow(info == 0, LAPACKSupport::ExcErrorCode("descinit_", info));
+    }
+  if (this->grid->mpi_process_is_active)
+    {
+      const int   ii = 1;
+      NumberType *loc_vals_A =
+        (this->values.size() > 0 && grid->mpi_process_is_active) ?
```
Comment (davydden) [inline source/lac/scalapack.cc]: `this->values.data()` should work regardless of size and being allocated.

File: source/lac/scalapack.cc [hunk 11]
```diff
@@ -263,6 +263,141 @@ ScaLAPACKMatrix<NumberType>::operator=(const FullMatrix<NumberType> &matrix)
...
+          &(this->values[0]) :
+          nullptr;
+      const NumberType *loc_vals_B =
+        mpi_process_is_active_B ? &(B(0, 0)) : nullptr;
+
+      // pgemr2d has to be called only for processes active on grid attached to
+      // matrix A
+      if (this->grid->mpi_process_is_active)
```
Comment (davydden) [inline source/lac/scalapack.cc]: you are already inside that `if`, so probably move the comment above.

File: source/lac/scalapack.cc [hunk 12]
```diff
@@ -298,6 +433,140 @@ ScaLAPACKMatrix<NumberType>::global_column(const unsigned int loc_column) const
...
+  // This grid will be used to copy to the distributed matrix to the serial
+  // matrix B using the ScaLAPACK routine pgemr2d.
+  MPI_Group group_A;
+  MPI_Comm_group(this->grid->mpi_communicator, &group_A);
+  const int              n = 1;
+  const std::vector<int> ranks(n, rank);
+  MPI_Group              group_B;
+  MPI_Group_incl(group_A, n, ranks.data(), &group_B);
+  MPI_Comm communicator_B;
+  MPI_Comm_create_group(MPI_COMM_WORLD, group_B, 0, &communicator_B);
```
Comment (davydden) [inline source/lac/scalapack.cc]: ditto

File: source/lac/scalapack.cc [hunk 13]
```diff
@@ -298,6 +433,140 @@ ScaLAPACKMatrix<NumberType>::global_column(const unsigned int loc_column) const
...
+                      &n_proc_rows_B,
+                      &n_proc_cols_B,
+                      &this_process_row_B,
+                      &this_process_column_B);
+      Assert(n_proc_rows_B * n_proc_cols_B == 1, ExcInternalError());
+      // the active process of grid B has to be process #rank of the
+      // communicator attached to A
+      Assert(this_mpi_process == rank, ExcInternalError());
+    }
+  bool mpi_process_is_active_B = false;
```
Comment (davydden) [inline source/lac/scalapack.cc]: ditto

File: source/lac/scalapack.cc [hunk 14]
```diff
@@ -298,6 +433,140 @@ ScaLAPACKMatrix<NumberType>::global_column(const unsigned int loc_column) const
...
+  if (this->grid->mpi_process_is_active)
+    {
+      const int         ii = 1;
+      const NumberType *loc_vals_A =
+        (this->values.size() > 0 && grid->mpi_process_is_active) ?
+          &(this->values[0]) :
+          nullptr;
+      NumberType *loc_vals_B = mpi_process_is_active_B ? &(B(0, 0)) : nullptr;
+
+      if (this->grid->mpi_process_is_active)
```
Comment (davydden) [inline source/lac/scalapack.cc]: you can remove that, you are already in this `if` branch.

File: source/lac/scalapack.cc [hunk 15]
```diff
@@ -298,6 +433,140 @@ ScaLAPACKMatrix<NumberType>::global_column(const unsigned int loc_column) const
...
+                &info);
+      AssertThrow(info == 0, LAPACKSupport::ExcErrorCode("descinit_", info));
+    }
+  // pgemr2d has to be called only for processes active on grid attached to
+  // matrix A
+  if (this->grid->mpi_process_is_active)
+    {
+      const int         ii = 1;
+      const NumberType *loc_vals_A =
+        (this->values.size() > 0 && grid->mpi_process_is_active) ?
```
Comment (davydden) [inline source/lac/scalapack.cc]: ditto

**Comment by BenBrands** [comment]
@davydden Any further suggestions?

File: source/lac/scalapack.cc (hunk 15 above)
Comment (davydden) [reply source/lac/scalapack.cc]: you can remove `grid->mpi_process_is_active` here

File: source/lac/scalapack.cc (hunk 10 above)
Comment (davydden) [reply source/lac/scalapack.cc]: you can remove `grid->mpi_process_is_active` here.

**Pull request merged**